### PR TITLE
[wip] ModelCheckpoint Callback save best_k_models score

### LINKED ISSUE #4911: ModelCheckpoint Callback save and restore extension
## 🚀 Feature

callbacks.ModelCheckpoint implements on_save_checkpoint and on_load_checkpoint to save and load it's state to/from a checkpoint. Save/Load all internal members to recover complete state from checkpoint. 

### Motivation

I use a ModelCheckpoint callback to save my Top-K (3) models during training. Sometimes I resume training at a certain point and want to train further epochs. But the loaded Callback has an empty best_k_models list, so this list is build up again. Older models are not tracked and I get three new models in same folder, and I do not know, which are the Top-3 ones. 
Therefore a full state of ModelCheckpoint would resolve this. 

### Pitch

Extend the ModelCheckpoint methods to export and load all internal states. I'm not sure if there is missing something, but here is a try:

    def on_save_checkpoint(self, trainer, pl_module) -> Dict[str, Any]:
        return {
            "best_model_score": self.best_model_score,
            "best_model_path": self.best_model_path,
            "best_k_models" : self.best_k_models,
        }

    def on_load_checkpoint(self, checkpointed_state: Dict[str, Any]):
        self.best_model_score = checkpointed_state["best_model_score"]
        self.best_model_path = checkpointed_state["best_model_path"]
        self.best_k_models = checkpointed_state["best_k_models"]

### Alternatives

Subclass the ModelCheckpoint to implement this, but I think this would complicate things.  

### Additional context

-/-

### PR DESCRIPTION
**Pull request by daandres**
## What does this PR do?

Saving the scores from best k models in ModelCHeckpoint Callback, to ensure correct state after restoring from checkpoint. 
Fixes #4911

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by daandres** [comment]
@carmocca @rohitgr7 
I tried the simple solution from issue, without other extensions. But I'm not seeing why I get a KeyError on a test which is not directly testing my lines. Could you help here? 
Thank you

**Comment by rohitgr7** [comment]
> @carmocca @rohitgr7
> I tried the simple solution from issue, without other extensions. But I'm not seeing why I get a KeyError on a test which is not directly testing my lines. Could you help here?
> Thank you

can you resolve the conflicts and rebase with the master branch. Would be easier to see the changes and issues you are facing here.
